### PR TITLE
First simple version

### PR DESCRIPTION
**Pull request by rianby64**
Let's begin to explore how to develop a simple table that works in websockets

### CODE REVIEW

File: web/src/Cell.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import React, { Component, ChangeEvent, FormEvent } from 'react'
+import { Form } from 'semantic-ui-react'
```
Comment (rianby64) [inline web/src/Cell.tsx]: @NovikovAntonY after a phone talk we decided to drop `semantic-ui-react` from this project. So, my suggestion is to replace any `semantic-ui-react` component with its html-plain corresponding tag.

File: web/src/Cell.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    this.props.onRedact(this.state.cell);
+    this.endRedact();
+  }
+
+  render() {
+    const { cell: { field, value }, edit } = this.state;
+    return edit ?
+    (
+      <Form onSubmit={this.submit}>
```
Comment (rianby64) [inline web/src/Cell.tsx]: `Form` should be replaced with a `Modal` component as you suggested.

File: web/src/Row.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    const { row } = this.props;
+    this.props.onDelete(row);
+  }
+
+  render() {
+    const { row, onRedact } = this.props;
+    return (
+      <Table.Row>
+        <Table.Cell collapsing>{row.ID}</Table.Cell>
+        <Table.Cell>
```
Comment (rianby64) [inline web/src/Row.tsx]: Yeap, replace them with `<td>`s

**Review by rianby64 — COMMENTED** [review]
This is part of our phone conversation we had.

File: web/src/Cell.tsx (hunk 1 above)
Comment (NovikovAntonY) [reply web/src/Cell.tsx]: @rianby64 ok, I will. Now i preparing for commit delete row feature. After this begin to remove semantic UI from project.

**Review by anovikau — COMMENTED** [review]
@rianby64 "What about if the list has millons of entries?". Yes, need scroll to bottom of page if you want to add new row. I think it nice decision, but not for this case, in my opinion. I think need to create buttons, near the table, it give us to advantages, for example, a clean table with unnecessary tags.

File: web/src/Row.tsx (hunk 3 above)
Comment (NovikovAntonY) [reply web/src/Row.tsx]: @rianby64 done

File: web/src/Cell.tsx (hunk 1 above)
Comment (NovikovAntonY) [reply web/src/Cell.tsx]: @rianby64 done

File: web/src/App.css [hunk 4]
```diff
@@ -1,32 +1,9 @@
...
-    transform: rotate(0deg);
-  }
-  to {
-    transform: rotate(360deg);
-  }
-}
+td {
+  border: 1px solid #cfcfcf;
+  padding: 0px 3px;
+}
```
Comment (rianby64) [inline web/src/App.css]: Please, for every file leave a `new line` at the end.

**Review by rianby64 — COMMENTED** [review]
Looks great!

**Comment by rianby64** [comment]
> @rianby64 "What about if the list has millons of entries?". Yes, need scroll to bottom of page if you want to add new row. I think it nice decision, but not for this case, in my opinion. I think need to create buttons, near the table, it give us to advantages, for example, a clean table with unnecessary tags.

The case with millions of rows should be avoided. Sounds nice the idea to have in the header an `add a new entry` button. Don't forget to static the header with that button at the top of the table, and if fire the action of creating a new entry then focus that new entry.

File: web/src/Table.tsx [hunk 5]
```diff
@@ -55,7 +55,15 @@ class TableArca extends Component {
...
             console.log(response, 'insert');
           } else if (response.Method == 'delete') {
-            console.log(response, 'delete');
+            const rowDeleted = response.Result;
+            this.setState((state: any) => {
+              const rows = state.rows.map((row: any) => {
+                if (row.ID !== rowDeleted.ID) {
+                  return row;
+                }
+              }).filter((row: any) => row !== undefined);
```
Comment (rianby64) [inline web/src/Table.tsx]: Next time you can use the filter to drop the _deleted_ entry. `map` looks redundant here.

**Review by rianby64 — COMMENTED** [review]
Looks great!

File: web/src/Table.tsx (hunk 5 above)
Comment (NovikovAntonY) [reply web/src/Table.tsx]: @rianby64 done

File: web/src/App.css (hunk 4 above)
Comment (NovikovAntonY) [reply web/src/App.css]: @rianby64 ok, sorry, corrected it.

File: web/src/Modal.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    this.props.request(data);
+    this.props.closeModal();
+  }
+
+  cancelCreating() {
+    this.props.closeModal();
+  }
+
+  changeNum1(event: ChangeEvent<HTMLInputElement>) {
```
Comment (NovikovAntonY) [inline web/src/Modal.tsx]: Don't think this moment the best decision, but best what i know, [react documentation](https://reactjs.org/docs/forms.html) don't have input multiply checks. In future i will replace that.

**Review by rianby64 — COMMENTED** [review]
Looks good! There are some things that I'd like to improve.

File: web/src/Modal.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    const newState = {
+      ...this.state,
+      Num1: event.target.value
```
Comment (NovikovAntonY) [inline web/src/Modal.tsx]: Don't think this moment the best decision, but best what i know, [react documentation](https://reactjs.org/docs/forms.html) don't have input multiply checks. In future i will replace that.

**Review by anovikau — COMMENTED** [review]
Remove old concept adds new row

File: web/src/Modal.tsx (hunk 7 above)
Comment (rianby64) [reply web/src/Modal.tsx]: Ok :+1:

**Review by rianby64 — COMMENTED** [review]
Can't imagine how you will handle the "create button" but I'm sure you'll champion it!

File: web/src/Table.tsx [hunk 8]
```diff
@@ -1,23 +1,16 @@
 import React, { Component } from 'react'
 import RowArca, { IRowArca } from './Row';
 
-class TableArca extends Component {
+class TableArca extends Component<any> {
```
[the remark this answers is not in the document]
Comment (NovikovAntonY) [reply web/src/Table.tsx]: @rianby64 yes, now for me put <any> means remove some  TS errors.

File: web/src/App.tsx [hunk 9]
```diff
@@ -1,12 +1,52 @@
...
+
+      return newState;
+    });
+  }
+
   render() {
     return (
       <div className="App">
-        <TableArca></TableArca>
+        <Buttom openModal={this.modalSwitcher} />
```
[the remark this answers is not in the document]
Comment (NovikovAntonY) [reply web/src/App.tsx]: @rianby64 done

File: web/src/Modal.tsx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    this.setState(newState);
+  }
+
+  render() {
+    return (
+      <div className="modal">
+        <form onSubmit={this.createRow} autoComplete="off">
+          <h2>Creating new row</h2>
+          <label htmlFor="num1">Num1</label>
+          <input id="num1" value={this.state.Num1} onChange={this.changeNum1} />
```
[the remark this answers is not in the document]
Comment (NovikovAntonY) [reply web/src/Modal.tsx]: done

File: web/src/Modal.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (NovikovAntonY) [reply web/src/Modal.tsx]: done

File: web/src/Modal.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    this.state = {
+      Num1: '',
+      Num2: '',
+      I: '',
+    }
+
+    this.createRow = this.createRow.bind(this);
+    this.cancelCreating = this.cancelCreating.bind(this);
+    this.changeNum1 = this.changeNum1.bind(this);
```
[the remark this answers is not in the document]
Comment (NovikovAntonY) [reply web/src/Modal.tsx]: done

File: web/src/App.tsx [hunk 12]
```diff
@@ -19,8 +20,22 @@ class App extends Component {
...
+      tables: [
+        {
+          source: 'ViewTable1',
+          rows: [],
+          fields: ['Num1', 'Num2', 'I'],
+          headers: ['ID', 'Num1', 'Num2', 'I'],
+        },
+        {
+          source: 'ViewTable2',
+          rows: [],
```
Comment (NovikovAntonY) [inline web/src/App.tsx]: now don't use, but in future I think state ` ViewTable1: [],` will be here

Comment (rianby64) [reply web/src/App.tsx]: Let's bring back the hard-coded values, please :smile:

**Comment by rianby64** [comment]
Close this example. Need the code just for reference.

**Pull request merged**